### PR TITLE
Added Webtoon.xyz presence

### LINKED ISSUE #3581: Webtoon | webtoon.xyz
**Prerequisites and essential questions** 
- [x] Is it a popular site?
- [x] Is the website older than 2 months? 
- [ ] Is the site locked to a specific country/region?
- [ ] Is the site a paid service? (e.g. Netflix, Hulu)
- [x] Does the website feature NSFW content? (e.g. porn, etc...)
- [ ] Are you a donator/patron?
- [x] Do you acknowledge that coding presences is completely voluntary and may take time for your service to be added regardless of priority?

**zee #0691?** 

**www.webtoon.xyz | Webtoon** 

**Just like the webtoon presence?** 

**Same as the webtoon icon**

### PR DESCRIPTION
**Pull request by MareGraphics**
Fully working webtoon.xyz presence.
resolve #3581
Screenshots:
image
image
image
image
image
image
All settings work too

### CODE REVIEW

File: websites/W/WebtoonXYZ/dist/metadata.json [hunk 1]
```diff
@@ -0,0 +1,42 @@
+{
+    "$schema": "https://schemas.premid.app/metadata/1.3",
+    "service": "Webtoon",
```
Comment (Bas950) [inline websites/W/WebtoonXYZ/dist/metadata.json:3]: service name and folder name must be the same

**Review by Bas950 — CHANGES_REQUESTED** [review]

File: websites/W/WebtoonXYZ/dist/metadata.json (hunk 1 above)
Comment (MareGraphics) [reply websites/W/WebtoonXYZ/dist/metadata.json:3]: Oh I didn't know that, I'll fix it

File: websites/W/WebtoonXYZ/presence.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  clientId: "836962986451140609"
+}),
+  browsingStamp = Math.floor(Date.now() / 1000);
+
+presence.on("UpdateData", async () => {
+  const v2icons = await presence.getSetting("v2icons"),
+        logo = await presence.getSetting("logo"),
+        buttons = await presence.getSetting("buttons"),
+        data: PresenceData = {
+          largeImageKey: (logo) == 0 ? "logo" : "logo-v2",
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:11]:
```suggestion
          largeImageKey: !logo ? "logo" : "logo-v2",
```
but i don't really see sense in two types of images :/

**Review by ririxidev — CHANGES_REQUESTED** [review]
and maybe just delete these `v2 images` they are unecessary...

**Comment by MareGraphics** [comment]
> and maybe just delete these `v2 images` they are unnecessary...

I don't mind deleting them, but I just took a bit more different approach with Webtoon and MangaSee allowing users to customize the look a tiny bit, I'll delete them when I'm home :D

File: websites/W/WebtoonXYZ/dist/metadata.json [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    },
+    "category": "anime",
+    "logo": "https://i.imgur.com/ZLoGNfR.png",
+    "thumbnail": "https://i.imgur.com/UkgngBO.png",
+    "version": "1.0.0",
+    "color": "#eb2e73",
+    "tags": [
+      "manhwa",
+      "webtoon"
+    ],
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/dist/metadata.json:16]:
```suggestion
    "tags": [
      "manhwa",
      "webtoon",
      "nsfw"
    ],
```
forgot, this site includes a lot of NSFW movies and ads on homepage so add nsfw tag too

**Review by ririxidev — CHANGES_REQUESTED** [review]

File: websites/W/WebtoonXYZ/dist/metadata.json (hunk 3 above)
Comment (ririxidev) [reply websites/W/WebtoonXYZ/dist/metadata.json:16]: Waiting.

**Review by Bas950 — DISMISSED** [review]

File: websites/W/WebtoonXYZ/dist/metadata.json [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        "title": "Show buttons",
+        "icon": "fas fa-compress-arrows-alt",
+        "value": true
+      },
+      {
+        "id": "v2icons",
+        "title": "V2 icons (circle)",
+        "icon": "fas fa-circle",
+        "value": false
+      },
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/dist/metadata.json:34]:
```suggestion
```

File: websites/W/WebtoonXYZ/presence.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    pathname = document.location.pathname;
+
+    if(pathname == "/" && window.location.search.substr(0,2) == "?s"){
+      const urlParams = new URLSearchParams(window.location.search),
+            nsfw = urlParams.get("adult"),
+            search = nsfw == "1" ? "nsfw" : nsfw == "0" ? "non nsfw" : urlParams.get("s"),
+            results = document.querySelector(".c-blog__heading > .h4").textContent.split(" ")[1];
+      data.details = "Searching:";
+      data.state = search + " 🔸 " + results +" results";
+      data.smallImageKey = (v2icons) ? "search-v2" : "search";
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:23]:
```suggestion
      data.smallImageKey = "search";
```

File: websites/W/WebtoonXYZ/presence.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      const results = document.querySelector(".c-blog__heading > .h4").textContent;
+      data.details = "Browsing all webtoons";
+      data.state = results;
+    }
+    else if(pathname.startsWith("/webtoon-genre/")){
+      const genre = document.querySelector(".item-title").textContent,
+            results = document.querySelector(".c-blog__heading > .h4").textContent;
+      data.details = "Browsing " + genre + " webtoons";
+      data.state = "📋 " +results; 
+      data.smallImageKey = (v2icons) ? "search-v2" : "search";
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:37]:
```suggestion
      data.smallImageKey = "search";
```

File: websites/W/WebtoonXYZ/presence.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    else if(pathname.startsWith("/read") && pathname.indexOf("/chapter") > 0){
+      const title = document.querySelector("#chapter-heading").textContent.split("-")[0],
+            chapter = document.querySelector("#chapter-heading").textContent.split("-")[1];
+      let progress = (document.documentElement.scrollTop) / (document.querySelector(".reading-content").scrollHeight - window.innerHeight)*100;
+      progress = Math.ceil(progress) > 100 ? 100 : Math.ceil(progress);
+      data.details = title;
+      data.state = "📖 " + chapter + " 🔸 " + progress + "%";
+      data.largeImageKey = title.includes("Solo Leveling") ? "solo" : (logo) == 0 ? "logo" : "logo-v2";
+      data.smallImageKey = (v2icons) ? "read-v2" : "read";
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:52]:
```suggestion
      data.smallImageKey = "read";
```

File: websites/W/WebtoonXYZ/presence.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+      const genre = document.querySelector(".item-title").textContent,
+            results = document.querySelector(".c-blog__heading > .h4").textContent;
+      data.details = "Browsing " + genre + " webtoons";
+      data.state = "📋 " +results; 
+      data.smallImageKey = (v2icons) ? "search-v2" : "search";
+    }
+    else if(pathname == "/completed-webtoons/"){
+      data.details = "Browsing:";
+      data.state = "Completed webtoons";
+      data.smallImageKey = (v2icons) ? "search-v2" : "search";
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:42]:
```suggestion
      data.smallImageKey = "search";
```

File: websites/W/WebtoonXYZ/presence.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+      data.state = "📖 " + chapter + " 🔸 " + progress + "%";
+      data.largeImageKey = title.includes("Solo Leveling") ? "solo" : (logo) == 0 ? "logo" : "logo-v2";
+      data.smallImageKey = (v2icons) ? "read-v2" : "read";
+      if (buttons) data.buttons = [{label: "Read Webtoon", url: window.location.href}];
+    }
+    else if(pathname.startsWith("/read")){
+      const title = document.querySelector(".post-title").textContent;
+      data.details = "Viewing:";
+      data.state = title;
+      data.smallImageKey = (v2icons) ? "view-v2" : "view";
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:59]:
```suggestion
      data.smallImageKey = "view";
```

File: websites/W/WebtoonXYZ/presence.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      data.largeImageKey = title.includes("Solo Leveling") ? "solo" : (logo) == 0 ? "logo" : "logo-v2";
+      if (buttons) data.buttons = [{label: "View Webtoon", url: window.location.href}];
+    }
+    else if(pathname == "/user-settings/"){
+      data.smallImageKey = (v2icons) ? "settings-v2" : "settings";
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:64]:
```suggestion
      data.smallImageKey = "settings";
```

File: websites/W/WebtoonXYZ/presence.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        data: PresenceData = {
+          largeImageKey: !logo ? "logo" : "logo-v2",
+          startTimestamp: browsingStamp
+        },
+    pathname = document.location.pathname;
+
+    if(pathname == "/" && window.location.search.substr(0,2) == "?s"){
+      const urlParams = new URLSearchParams(window.location.search),
+            nsfw = urlParams.get("adult"),
+            search = nsfw == "1" ? "nsfw" : nsfw == "0" ? "non nsfw" : urlParams.get("s"),
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:19]:
```suggestion
            search = nsfw === "1" ? "nsfw" : nsfw === "0" ? "non nsfw" : urlParams.get("s"),
```

**Review by ririxidev — CHANGES_REQUESTED** [review]
made your work easier :d

File: websites/W/WebtoonXYZ/presence.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+            results = document.querySelector(".c-blog__heading > .h4").textContent.split(" ")[1];
+      data.details = "Searching:";
+      data.state = search + " 🔸 " + results +" results";
+      data.smallImageKey = (v2icons) ? "search-v2" : "search";
+    }    
+    else if(pathname == "/")
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:25]:
```suggestion
    else if(pathname === "/")
```

File: websites/W/WebtoonXYZ/presence.ts [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+      data.state = results;
+    }
+    else if(pathname.startsWith("/webtoon-genre/")){
+      const genre = document.querySelector(".item-title").textContent,
+            results = document.querySelector(".c-blog__heading > .h4").textContent;
+      data.details = "Browsing " + genre + " webtoons";
+      data.state = "📋 " +results; 
+      data.smallImageKey = (v2icons) ? "search-v2" : "search";
+    }
+    else if(pathname == "/completed-webtoons/"){
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:39]:
```suggestion
    else if(pathname === "/completed-webtoons/"){
```

File: websites/W/WebtoonXYZ/presence.ts [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+    }
+    else if(pathname.startsWith("/read")){
+      const title = document.querySelector(".post-title").textContent;
+      data.details = "Viewing:";
+      data.state = title;
+      data.smallImageKey = (v2icons) ? "view-v2" : "view";
+      data.largeImageKey = title.includes("Solo Leveling") ? "solo" : (logo) == 0 ? "logo" : "logo-v2";
+      if (buttons) data.buttons = [{label: "View Webtoon", url: window.location.href}];
+    }
+    else if(pathname == "/user-settings/"){
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:63]:
```suggestion
    else if(pathname === "/user-settings/"){
```

File: websites/W/WebtoonXYZ/presence.ts [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+  const v2icons = await presence.getSetting("v2icons"),
+        logo = await presence.getSetting("logo"),
+        buttons = await presence.getSetting("buttons"),
+        data: PresenceData = {
+          largeImageKey: !logo ? "logo" : "logo-v2",
+          startTimestamp: browsingStamp
+        },
+    pathname = document.location.pathname;
+
+    if(pathname == "/" && window.location.search.substr(0,2) == "?s"){
```
Comment (ririxidev) [inline websites/W/WebtoonXYZ/presence.ts:16]:
```suggestion
    if(pathname === "/" && window.location.search.substr(0,2) == "?s"){
```

**Review by ririxidev — CHANGES_REQUESTED** [review]
forgot about these stuff, please always check type so instead `==` use `===`

**Comment by MareGraphics** [comment]
> made your work easier :d

Well, thanks a lot xD

**Comment by MareGraphics** [comment]
> forgot about these stuff, please always check type so instead `==` use `===`

Oh, I'll pay attention to it on my next presences

**Comment by ririxidev** [comment]
but because of that, you need to squash your commits :3

**Comment by MareGraphics** [comment]
Well to gain this I lose that xD

**Comment by ririxidev** [comment]
Waiting, and it is necessary step if you want us to accept your pull request.

**Comment by ririxidev** [comment]
well - you don't need now kek

**Pull request merged**